File: python_programs_and_containers/system_control/file_loader/test_files/kb_go/postgres/kb_construct/kb_construct_module/KnowledgeBaseManager.go
package kb_construct_module

import (
	"database/sql"
	"encoding/json"
	"fmt"
	//"log"
	//"os"
	//"strings"

	_ "github.com/lib/pq"
)

// KnowledgeBaseManager manages knowledge base operations
type KnowledgeBaseManager struct {
	conn      *sql.DB
	tableName string
}

// ConnectionParams holds database connection parameters
type ConnectionParams struct {
	Host     string
	Database string
	User     string
	Password string
	Port     int
}

// NewKnowledgeBaseManager creates a new instance of KnowledgeBaseManager
func NewKnowledgeBaseManager(tableName string, connParams ConnectionParams) (*KnowledgeBaseManager, error) {
	connStr := fmt.Sprintf("host=%s port=%d user=%s password=%s dbname=%s sslmode=disable",
		connParams.Host, connParams.Port, connParams.User, connParams.Password, connParams.Database)

	db, err := sql.Open("postgres", connStr)
	if err != nil {
		return nil, fmt.Errorf("error connecting to database: %w", err)
	}

	// Test the connection
	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("error pinging database: %w", err)
	}

	kb := &KnowledgeBaseManager{
		conn:      db,
		tableName: tableName,
	}

	// Enable ltree extension
	if _, err := kb.conn.Exec("CREATE EXTENSION IF NOT EXISTS ltree;"); err != nil {
		return nil, fmt.Errorf("error creating ltree extension: %w", err)
	}

	// Create tables
	if err := kb.createTables(); err != nil {
		return nil, fmt.Errorf("error creating tables: %w", err)
	}

	return kb, nil
}

// Disconnect closes the database connection
func (kb *KnowledgeBaseManager) Disconnect() error {
	if kb.conn != nil {
		return kb.conn.Close()
	}
	return nil
}

// deleteTable deletes a specified table
func (kb *KnowledgeBaseManager) deleteTable(tableName string, schema string) error {
	query := fmt.Sprintf("DROP TABLE IF EXISTS %s.%s CASCADE;", schema, tableName)
	_, err := kb.conn.Exec(query)
	if err != nil {
		return fmt.Errorf("error deleting table %s.%s: %w", schema, tableName, err)
	}
	return nil
}

// createTables creates all necessary tables
func (kb *KnowledgeBaseManager) createTables() error {
	// Delete existing tables
	tables := []string{
		kb.tableName,
		kb.tableName + "_info",
		kb.tableName + "_link",
		kb.tableName + "_link_mount",
	}
	for _, table := range tables {
		//fmt.Println("deleting table", table)
		if err := kb.deleteTable(table, "public"); err != nil {
			return err
		}
	}
	

	// Create main knowledge base table
	kbTableQuery := fmt.Sprintf(`
		CREATE TABLE %s (
			id SERIAL PRIMARY KEY,
			knowledge_base VARCHAR NOT NULL,
			label VARCHAR NOT NULL,
			name VARCHAR NOT NULL,
			properties JSON,
			data JSON,
			has_link BOOLEAN DEFAULT FALSE,
			has_link_mount BOOLEAN DEFAULT FALSE,
			path LTREE UNIQUE
		)`, kb.tableName)

	if _, err := kb.conn.Exec(kbTableQuery); err != nil {
		return fmt.Errorf("error creating knowledge base table: %w", err)
	}

	// Create info table
	infoTableQuery := fmt.Sprintf(`
		CREATE TABLE %s_info (
			id SERIAL PRIMARY KEY,
			knowledge_base VARCHAR NOT NULL UNIQUE,
			description VARCHAR
		)`, kb.tableName)

	if _, err := kb.conn.Exec(infoTableQuery); err != nil {
		return fmt.Errorf("error creating info table: %w", err)
	}

	// Create link table
	linkTableQuery := fmt.Sprintf(`
		CREATE TABLE %s_link (
			id SERIAL PRIMARY KEY,
			link_name VARCHAR NOT NULL,
			parent_node_kb VARCHAR NOT NULL,
			parent_path LTREE NOT NULL,
			created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
			UNIQUE(link_name, parent_node_kb, parent_path)
		)`, kb.tableName)

	if _, err := kb.conn.Exec(linkTableQuery); err != nil {
		return fmt.Errorf("error creating link table: %w", err)
	}

	// Create link mount table
	linkMountTableQuery := fmt.Sprintf(`
		CREATE TABLE %s_link_mount (
			id SERIAL PRIMARY KEY,
			link_name VARCHAR NOT NULL UNIQUE,
			knowledge_base VARCHAR NOT NULL,
			mount_path LTREE NOT NULL,
			description VARCHAR,
			created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
			UNIQUE(knowledge_base, mount_path)
		)`, kb.tableName)

	if _, err := kb.conn.Exec(linkMountTableQuery); err != nil {
		return fmt.Errorf("error creating link mount table: %w", err)
	}

	// Create all indexes
	return kb.createIndexes()
}

// createIndexes creates all necessary indexes
func (kb *KnowledgeBaseManager) createIndexes() error {
	indexes := []string{
		// Main table indexes
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_kb ON %s (knowledge_base)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_path ON %s USING GIST (path)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_label ON %s (label)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_name ON %s (name)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_has_link ON %s (has_link)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_has_link_mount ON %s (has_link_mount)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_kb_path ON %s (knowledge_base, path)", kb.tableName, kb.tableName),

		// Info table indexes
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_info_kb ON %s_info (knowledge_base)", kb.tableName, kb.tableName),

		// Link table indexes
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_link_name ON %s_link (link_name)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_link_parent_kb ON %s_link (parent_node_kb)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_link_parent_path ON %s_link USING GIST (parent_path)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_link_created ON %s_link (created_at)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_link_composite ON %s_link (link_name, parent_node_kb)", kb.tableName, kb.tableName),

		// Mount table indexes
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_mount_link_name ON %s_link_mount (link_name)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_mount_kb ON %s_link_mount (knowledge_base)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_mount_path ON %s_link_mount USING GIST (mount_path)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_mount_created ON %s_link_mount (created_at)", kb.tableName, kb.tableName),
		fmt.Sprintf("CREATE INDEX IF NOT EXISTS idx_%s_mount_composite ON %s_link_mount (knowledge_base, mount_path)", kb.tableName, kb.tableName),
	}

	for _, indexQuery := range indexes {
		if _, err := kb.conn.Exec(indexQuery); err != nil {
			return fmt.Errorf("error creating index: %w", err)
		}
	}

	return nil
}

// AddKB adds a knowledge base entry to the information table
func (kb *KnowledgeBaseManager) AddKB(kbName string, description string) error {
	infoTable := kb.tableName + "_info"
	query := fmt.Sprintf(`
		INSERT INTO %s (knowledge_base, description)
		VALUES ($1, $2)
		ON CONFLICT (knowledge_base) DO NOTHING`, infoTable)

	_, err := kb.conn.Exec(query, kbName, description)
	if err != nil {
		return fmt.Errorf("error adding knowledge base: %w", err)
	}

	return nil
}

// AddNode adds a node to the knowledge base
func (kb *KnowledgeBaseManager) AddNode(kbName, label, name string, properties, data map[string]interface{}, path string) error {
	// Check if kb_name exists in info table
	infoTable := kb.tableName + "_info"
	checkQuery := fmt.Sprintf("SELECT 1 FROM %s WHERE knowledge_base = $1", infoTable)

	var exists int
	err := kb.conn.QueryRow(checkQuery, kbName).Scan(&exists)
	if err == sql.ErrNoRows {
		return fmt.Errorf("knowledge base '%s' not found in info table", kbName)
	} else if err != nil {
		return fmt.Errorf("error checking knowledge base: %w", err)
	}

	// Convert maps to JSON
	var propertiesJSON, dataJSON []byte
	if properties != nil {
		propertiesJSON, err = json.Marshal(properties)
		if err != nil {
			return fmt.Errorf("error marshaling properties: %w", err)
		}
	}
	if data != nil {
		dataJSON, err = json.Marshal(data)
		if err != nil {
			return fmt.Errorf("error marshaling data: %w", err)
		}
	}

	// Insert node
	insertQuery := fmt.Sprintf(`
		INSERT INTO %s (knowledge_base, label, name, properties, data, has_link, path)
		VALUES ($1, $2, $3, $4, $5, $6, $7)`, kb.tableName)

	_, err = kb.conn.Exec(insertQuery, kbName, label, name, propertiesJSON, dataJSON, false, path)
	if err != nil {
		return fmt.Errorf("error adding node: %w", err)
	}

	return nil
}

// AddLink adds a link between nodes
func (kb *KnowledgeBaseManager) AddLink(parentKB, parentPath, linkName string) error {
	// Check if parent knowledge base exists
	infoTable := kb.tableName + "_info"
	kbCheckQuery := fmt.Sprintf("SELECT knowledge_base FROM %s WHERE knowledge_base = $1", infoTable)

	var foundKB string
	err := kb.conn.QueryRow(kbCheckQuery, parentKB).Scan(&foundKB)
	if err == sql.ErrNoRows {
		return fmt.Errorf("parent knowledge base '%s' not found", parentKB)
	} else if err != nil {
		return fmt.Errorf("error checking knowledge base: %w", err)
	}

	// Check if parent node exists
	nodeCheckQuery := fmt.Sprintf("SELECT path FROM %s WHERE path = $1", kb.tableName)
	var foundPath string
	err = kb.conn.QueryRow(nodeCheckQuery, parentPath).Scan(&foundPath)
	if err == sql.ErrNoRows {
		return fmt.Errorf("parent node with path '%s' not found", parentPath)
	} else if err != nil {
		return fmt.Errorf("error checking node: %w", err)
	}

	// Check if link name already exists in link_mount table
	linkTable := kb.tableName + "_link"
	linkNameExistsQuery := fmt.Sprintf("SELECT link_name FROM %s WHERE link_name = $1", linkTable)
	var existingLinkName string
	err = kb.conn.QueryRow(linkNameExistsQuery, linkName).Scan(&existingLinkName)
	if err != sql.ErrNoRows {
		return fmt.Errorf("link name '%s' already exists in link_mount table", linkName)
	}

	// Begin transaction
	tx, err := kb.conn.Begin()
	if err != nil {
		return fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	
	linkInsertQuery := fmt.Sprintf(`
		INSERT INTO %s (parent_node_kb, parent_path, link_name)
		VALUES ($1, $2, $3)`, linkTable)

	_, err = tx.Exec(linkInsertQuery, parentKB, parentPath, linkName)
	if err != nil {
		return fmt.Errorf("error inserting link: %w", err)
	}

	// Update has_link flag
	updateQuery := fmt.Sprintf("UPDATE %s SET has_link = TRUE WHERE path = $1", kb.tableName)
	_, err = tx.Exec(updateQuery, parentPath)
	if err != nil {
		return fmt.Errorf("error updating has_link flag: %w", err)
	}

	// Commit transaction
	if err = tx.Commit(); err != nil {
		return fmt.Errorf("error committing transaction: %w", err)
	}

	return nil
}

// AddLinkMount adds a link mount
func (kb *KnowledgeBaseManager) AddLinkMount(knowledgeBase, path, linkMountName, description string) (string, string, error) {
	// Verify that knowledge_base exists in info table
	infoCheckQuery := fmt.Sprintf("SELECT knowledge_base FROM %s_info WHERE knowledge_base = $1", kb.tableName)
	var foundKB string
	err := kb.conn.QueryRow(infoCheckQuery, knowledgeBase).Scan(&foundKB)
	if err == sql.ErrNoRows {
		return "", "", fmt.Errorf("knowledge base '%s' does not exist in info table", knowledgeBase)
	} else if err != nil {
		return "", "", fmt.Errorf("error checking knowledge base: %w", err)
	}

	// Verify that the path exists for the given knowledge base
	pathCheckQuery := fmt.Sprintf("SELECT id FROM %s WHERE knowledge_base = $1 AND path = $2", kb.tableName)
	var nodeID int
	err = kb.conn.QueryRow(pathCheckQuery, knowledgeBase, path).Scan(&nodeID)
	if err == sql.ErrNoRows {
		return "", "", fmt.Errorf("path '%s' does not exist for knowledge base '%s'", path, knowledgeBase)
	} else if err != nil {
		return "", "", fmt.Errorf("error checking path: %w", err)
	}

	// Verify that link_name does not already exist in link_mount table
	linkNameExistsQuery := fmt.Sprintf("SELECT link_name FROM %s_link_mount WHERE link_name = $1", kb.tableName)
	var existingLinkName string
	err = kb.conn.QueryRow(linkNameExistsQuery, linkMountName).Scan(&existingLinkName)
	if err != sql.ErrNoRows {
		return "", "", fmt.Errorf("link name '%s' already exists in link_mount table", linkMountName)
	}

	// Begin transaction
	tx, err := kb.conn.Begin()
	if err != nil {
		return "", "", fmt.Errorf("error beginning transaction: %w", err)
	}
	defer tx.Rollback()

	// Insert record in link_mount table
	insertLinkMountQuery := fmt.Sprintf(`
		INSERT INTO %s_link_mount (link_name, knowledge_base, mount_path, description)
		VALUES ($1, $2, $3, $4)`, kb.tableName)

	result, err := tx.Exec(insertLinkMountQuery, linkMountName, knowledgeBase, path, description)
	if err != nil {
		return "", "", fmt.Errorf("error inserting link mount: %w", err)
	}

	rowsAffected, err := result.RowsAffected()
	if err != nil {
		return "", "", fmt.Errorf("error getting rows affected: %w", err)
	}
	if rowsAffected == 0 {
		return "", "", fmt.Errorf("failed to insert record with link_name '%s', knowledge_base '%s', path '%s' into link_mount table", linkMountName, knowledgeBase, path)
	}

	// Update has_link_mount flag
	updateQuery := fmt.Sprintf(`
		UPDATE %s SET has_link_mount = TRUE 
		WHERE knowledge_base = $1 AND path = $2`, kb.tableName)

	result, err = tx.Exec(updateQuery, knowledgeBase, path)
	if err != nil {
		return "", "", fmt.Errorf("error updating has_link_mount flag: %w", err)
	}

	rowsAffected, err = result.RowsAffected()
	if err != nil {
		return "", "", fmt.Errorf("error getting rows affected: %w", err)
	}
	if rowsAffected == 0 {
		return "", "", fmt.Errorf("no rows were updated for knowledge_base '%s' and path '%s'", knowledgeBase, path)
	}

	// Commit transaction
	if err = tx.Commit(); err != nil {
		return "", "", fmt.Errorf("error committing transaction: %w", err)
	}

	return knowledgeBase, path, nil
}

/*
func main() {
	// Get password from user
	fmt.Print("Enter PostgreSQL password: ")
	var password string
	//fmt.Scanln(&password)
	password := os.Getenv("POSTGRES_PASSWORD")
	if password == "" {
		fmt.Println("password is empty")
		os.Exit(1)
	}

	// Database connection parameters
	connParams := ConnectionParams{
		Host:     "localhost",
		Database: "knowledge_base",
		User:     "gedgar",
		Password: password,
		Port:     5432,
	}

	// Create knowledge base manager
	kbManager, err := NewKnowledgeBaseManager("knowledge_base", connParams)
	if err != nil {
		log.Fatal("Error creating knowledge base manager:", err)
	}
	defer kbManager.Disconnect()

	fmt.Println("starting unit test")

	// Add knowledge bases
	if err := kbManager.AddKB("kb1", "First knowledge base"); err != nil {
		log.Fatal("Error adding kb1:", err)
	}
	if err := kbManager.AddKB("kb2", "Second knowledge base"); err != nil {
		log.Fatal("Error adding kb2:", err)
	}

	// Add nodes
	properties1 := map[string]interface{}{"age": 30}
	data1 := map[string]interface{}{"email": "john@example.com"}
	if err := kbManager.AddNode("kb1", "person", "John Doe", properties1, data1, "people.john"); err != nil {
		log.Fatal("Error adding John Doe node:", err)
	}

	properties2 := map[string]interface{}{"age": 25}
	data2 := map[string]interface{}{"email": "jane@example.com"}
	if err := kbManager.AddNode("kb2", "person", "Jane Smith", properties2, data2, "people.jane"); err != nil {
		log.Fatal("Error adding Jane Smith node:", err)
	}

	// Add link mount
	_, _, err = kbManager.AddLinkMount("kb1", "people.john", "link1", "link1 description")
	if err != nil {
		log.Fatal("Error adding link mount:", err)
	}
	

	// Add link
	if err := kbManager.AddLink("kb1", "people.john", "link1"); err != nil {
		log.Fatal("Error adding link:", err)
	}

	fmt.Println("ending unit test")
	os.Exit(1)
}

*/

File: luajit_programs_and_containers/building_blocks/knowledge_base/postgres/construct_kb/knowledge_base_manager.lua
--[[
  knowledge_base_manager.lua
  
  LuaJIT PostgreSQL knowledge base manager using luadbi (libpq wrapper).
  
  Dependencies:
    - luadbi + luadbi-postgresql  (C wrapper around libpq)
    - dkjson                      (pure Lua JSON)
  
  Install:
    sudo apt install libpq-dev
    luarocks --lua-version 5.1 install luadbi
    luarocks --lua-version 5.1 install luadbi-postgresql
    luarocks --lua-version 5.1 install dkjson
  
  Usage:
    local KBManager = require("knowledge_base_manager")
    local mgr = KBManager.new("knowledge_base", conn_params)
]]

local DBI = require("DBI")
local json = require("dkjson")

local KnowledgeBaseManager = {}
KnowledgeBaseManager.__index = KnowledgeBaseManager

--- Escape a SQL identifier (table/column name).
local function quote_ident(name)
  return '"' .. name:gsub('"', '""') .. '"'
end

--- Escape a SQL literal string value.
local function quote_literal(val)
  if val == nil then return "NULL" end
  return "'" .. tostring(val):gsub("'", "''") .. "'"
end

---------------------------------------------------------------------------
-- Constructor
---------------------------------------------------------------------------

function KnowledgeBaseManager.new(table_name, connection_params, upload_flag)
  assert(type(table_name) == "string", "table_name must be a string")
  assert(type(connection_params) == "table", "connection_params must be a table")

  local self = setmetatable({}, KnowledgeBaseManager)
  self.table_name = table_name
  self.connection_params = connection_params
  self.upload_flag = upload_flag or false
  self.conn = nil

  self:_connect()
  if not self.upload_flag then
    self:_create_tables()
  end

  return self
end

---------------------------------------------------------------------------
-- Connection
---------------------------------------------------------------------------

function KnowledgeBaseManager:_connect()
  local params = self.connection_params
  local dbname = params.database or params.dbname

  local conn, err = DBI.Connect("PostgreSQL",
    dbname,
    params.user,
    params.password,
    params.host or "127.0.0.1",
    tostring(params.port or 5432))

  if not conn then
    error("Error connecting to database: " .. tostring(err))
  end

  self.conn = conn
  self.conn:autocommit(true)

  -- Enable ltree extension
  self:_exec("CREATE EXTENSION IF NOT EXISTS ltree;")
end

function KnowledgeBaseManager:disconnect()
  if self.conn then
    self.conn:close()
    self.conn = nil
  end
end

---------------------------------------------------------------------------
-- SQL execution helpers
---------------------------------------------------------------------------

--- Execute a SQL statement (no result set expected).
function KnowledgeBaseManager:_exec(sql_str)
  local stmt, err = self.conn:prepare(sql_str)
  if not stmt then
    error("SQL prepare error: " .. tostring(err) .. "\nQuery: " .. sql_str)
  end
  local ok, exec_err = stmt:execute()
  if not ok then
    error("SQL execute error: " .. tostring(exec_err) .. "\nQuery: " .. sql_str)
  end
  stmt:close()
end

--- Execute a query and return all rows as an array of named-column tables.
function KnowledgeBaseManager:_query(sql_str)
  local stmt, err = self.conn:prepare(sql_str)
  if not stmt then
    error("SQL prepare error: " .. tostring(err) .. "\nQuery: " .. sql_str)
  end
  local ok, exec_err = stmt:execute()
  if not ok then
    error("SQL execute error: " .. tostring(exec_err) .. "\nQuery: " .. sql_str)
  end

  local rows = {}
  local row = stmt:fetch(true)  -- true = named columns
  while row do
    rows[#rows + 1] = row
    row = stmt:fetch(true)
  end
  stmt:close()
  return rows
end

---------------------------------------------------------------------------
-- Table management
---------------------------------------------------------------------------

function KnowledgeBaseManager:_delete_table(tbl_name, schema)
  schema = schema or "public"
  self:_exec(string.format("DROP TABLE IF EXISTS %s.%s CASCADE;",
    quote_ident(schema), quote_ident(tbl_name)))
end

--- Terminate any other connections to this database that are idle in
--- transaction. They hold locks from a previously crashed/aborted build
--- and would block our DROP TABLE indefinitely otherwise.
function KnowledgeBaseManager:_terminate_stale_locks()
  -- Don't error if this fails — best-effort cleanup. Wrapped in pcall
  -- so a missing privilege doesn't kill the whole build.
  pcall(function()
    local rows = self:_query([[
      SELECT pg_terminate_backend(pid) AS killed, pid
      FROM pg_stat_activity
      WHERE datname = current_database()
        AND pid <> pg_backend_pid()
        AND state IN ('idle in transaction', 'idle in transaction (aborted)');
    ]])
    if rows and #rows > 0 then
      io.stderr:write(string.format(
        "knowledge_base_manager: terminated %d stale idle-in-transaction connection(s)\n",
        #rows))
    end
  end)
end

function KnowledgeBaseManager:_create_tables()
  local tn = self.table_name

  -- Clean up any stale connections that might be holding locks on the
  -- tables we're about to drop. Without this, a previously crashed build
  -- can leave an idle-in-transaction connection that blocks the DROP
  -- forever. Then set a lock_timeout so even if something we couldn't
  -- terminate is holding the lock, we fail fast with a clear error
  -- instead of hanging.
  self:_terminate_stale_locks()
  pcall(function() self:_exec("SET lock_timeout = '10s';") end)

  self:_delete_table(tn)
  self:_delete_table(tn .. "_info")
  self:_delete_table(tn .. "_link")
  self:_delete_table(tn .. "_link_mount")

  self:_exec(string.format([[
    CREATE TABLE %s (
      id SERIAL PRIMARY KEY,
      knowledge_base VARCHAR NOT NULL,
      label VARCHAR NOT NULL,
      name VARCHAR NOT NULL,
      properties JSON,
      data JSON,
      has_link BOOLEAN DEFAULT FALSE,
      has_link_mount BOOLEAN DEFAULT FALSE,
      path LTREE UNIQUE
    )
  ]], quote_ident(tn)))

  self:_exec(string.format([[
    CREATE TABLE %s (
      id SERIAL PRIMARY KEY,
      knowledge_base VARCHAR NOT NULL UNIQUE,
      description VARCHAR
    )
  ]], quote_ident(tn .. "_info")))

  self:_exec(string.format([[
    CREATE TABLE %s (
      id SERIAL PRIMARY KEY,
      link_name VARCHAR NOT NULL,
      parent_node_kb VARCHAR NOT NULL,
      parent_path LTREE NOT NULL,
      created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
      UNIQUE(link_name, parent_node_kb, parent_path)
    )
  ]], quote_ident(tn .. "_link")))

  self:_exec(string.format([[
    CREATE TABLE %s (
      id SERIAL PRIMARY KEY,
      link_name VARCHAR NOT NULL UNIQUE,
      knowledge_base VARCHAR NOT NULL,
      mount_path LTREE NOT NULL,
      description VARCHAR,
      created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
      UNIQUE(knowledge_base, mount_path)
    )
  ]], quote_ident(tn .. "_link_mount")))

  -- Indexes
  local indexes = {
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (knowledge_base)",
      quote_ident("idx_"..tn.."_kb"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s USING GIST (path)",
      quote_ident("idx_"..tn.."_path"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (label)",
      quote_ident("idx_"..tn.."_label"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (name)",
      quote_ident("idx_"..tn.."_name"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (has_link)",
      quote_ident("idx_"..tn.."_has_link"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (has_link_mount)",
      quote_ident("idx_"..tn.."_has_link_mount"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (knowledge_base, path)",
      quote_ident("idx_"..tn.."_kb_path"), quote_ident(tn)),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (knowledge_base)",
      quote_ident("idx_"..tn.."_info_kb"), quote_ident(tn.."_info")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (link_name)",
      quote_ident("idx_"..tn.."_link_name"), quote_ident(tn.."_link")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (parent_node_kb)",
      quote_ident("idx_"..tn.."_link_parent_kb"), quote_ident(tn.."_link")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s USING GIST (parent_path)",
      quote_ident("idx_"..tn.."_link_parent_path"), quote_ident(tn.."_link")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (created_at)",
      quote_ident("idx_"..tn.."_link_created"), quote_ident(tn.."_link")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (link_name, parent_node_kb)",
      quote_ident("idx_"..tn.."_link_composite"), quote_ident(tn.."_link")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (link_name)",
      quote_ident("idx_"..tn.."_mount_link_name"), quote_ident(tn.."_link_mount")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (knowledge_base)",
      quote_ident("idx_"..tn.."_mount_kb"), quote_ident(tn.."_link_mount")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s USING GIST (mount_path)",
      quote_ident("idx_"..tn.."_mount_path"), quote_ident(tn.."_link_mount")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (created_at)",
      quote_ident("idx_"..tn.."_mount_created"), quote_ident(tn.."_link_mount")),
    string.format("CREATE INDEX IF NOT EXISTS %s ON %s (knowledge_base, mount_path)",
      quote_ident("idx_"..tn.."_mount_composite"), quote_ident(tn.."_link_mount")),
  }

  for _, idx_sql in ipairs(indexes) do
    self:_exec(idx_sql)
  end

  -- Restore default lock_timeout for the rest of this session
  pcall(function() self:_exec("SET lock_timeout = 0;") end)
end

---------------------------------------------------------------------------
-- Public API
---------------------------------------------------------------------------

function KnowledgeBaseManager:add_kb(kb_name, description)
  assert(type(kb_name) == "string", "kb_name must be a string")
  assert(type(description) == "string", "description must be a string")

  local info_table = quote_ident(self.table_name .. "_info")
  self:_exec(string.format([[
    INSERT INTO %s (knowledge_base, description)
    VALUES (%s, %s)
    ON CONFLICT (knowledge_base) DO NOTHING
  ]], info_table, quote_literal(kb_name), quote_literal(description)))
end

function KnowledgeBaseManager:add_node(kb_name, label, name, properties, data, path)
  assert(type(kb_name) == "string", "kb_name must be a string")
  assert(type(label) == "string", "label must be a string")
  assert(type(name) == "string", "name must be a string")
  assert(type(path) == "string", "path must be a string")
  assert(type(properties) == "table", "properties must be a table")
  assert(type(data) == "table", "data must be a table")

  local info_table = quote_ident(self.table_name .. "_info")
  local check = self:_query(string.format(
    "SELECT 1 FROM %s WHERE knowledge_base = %s",
    info_table, quote_literal(kb_name)))

  if #check == 0 then
    error("Knowledge base '" .. kb_name .. "' not found in info table")
  end

  local props_json = json.encode(properties)
  local data_json  = json.encode(data)

  local main_table = quote_ident(self.table_name)
  self:_exec(string.format([[
    INSERT INTO %s (knowledge_base, label, name, properties, data, has_link, path)
    VALUES (%s, %s, %s, %s, %s, FALSE, %s)
  ]], main_table,
      quote_literal(kb_name),
      quote_literal(label),
      quote_literal(name),
      quote_literal(props_json),
      quote_literal(data_json),
      quote_literal(path)))
end

function KnowledgeBaseManager:add_link(parent_kb, parent_path, link_name)
  assert(type(parent_kb) == "string", "parent_kb must be a string")
  assert(type(parent_path) == "string", "parent_path must be a string")
  assert(type(link_name) == "string", "link_name must be a string")

  local info_table = quote_ident(self.table_name .. "_info")
  local found = self:_query(string.format(
    "SELECT knowledge_base FROM %s WHERE knowledge_base = %s",
    info_table, quote_literal(parent_kb)))

  if #found == 0 then
    error("Parent knowledge base '" .. parent_kb .. "' not found")
  end

  local main_table = quote_ident(self.table_name)
  local paths = self:_query(string.format(
    "SELECT path FROM %s WHERE path = %s",
    main_table, quote_literal(parent_path)))

  if #paths == 0 then
    error("Parent node with path '" .. parent_path .. "' not found")
  end

  local mount_table = quote_ident(self.table_name .. "_link_mount")
  local mount_check = self:_query(string.format(
    "SELECT link_name FROM %s WHERE link_name = %s",
    mount_table, quote_literal(link_name)))

  if #mount_check == 0 then
    error("Link name '" .. link_name .. "' not found in link_mount table")
  end

  local link_table = quote_ident(self.table_name .. "_link")
  self:_exec(string.format([[
    INSERT INTO %s (parent_node_kb, parent_path, link_name)
    VALUES (%s, %s, %s)
  ]], link_table,
      quote_literal(parent_kb),
      quote_literal(parent_path),
      quote_literal(link_name)))

  self:_exec(string.format(
    "UPDATE %s SET has_link = TRUE WHERE path = %s",
    main_table, quote_literal(parent_path)))
end

function KnowledgeBaseManager:add_link_mount(knowledge_base, path, link_mount_name, description)
  description = description or ""
  assert(type(knowledge_base) == "string", "knowledge_base must be a string")
  assert(type(path) == "string", "path must be a string")
  assert(type(link_mount_name) == "string", "link_mount_name must be a string")
  assert(type(description) == "string", "description must be a string")

  local info_table  = quote_ident(self.table_name .. "_info")
  local main_table  = quote_ident(self.table_name)
  local mount_table = quote_ident(self.table_name .. "_link_mount")

  local kb_check = self:_query(string.format(
    "SELECT knowledge_base FROM %s WHERE knowledge_base = %s",
    info_table, quote_literal(knowledge_base)))

  if #kb_check == 0 then
    error("Knowledge base '" .. knowledge_base .. "' does not exist in info table")
  end

  local path_check = self:_query(string.format(
    "SELECT id FROM %s WHERE knowledge_base = %s AND path = %s",
    main_table, quote_literal(knowledge_base), quote_literal(path)))

  if #path_check == 0 then
    error("Path '" .. path .. "' does not exist for knowledge base '" .. knowledge_base .. "'")
  end

  local dup_check = self:_query(string.format(
    "SELECT link_name FROM %s WHERE link_name = %s",
    mount_table, quote_literal(link_mount_name)))

  if #dup_check > 0 then
    error("Link name '" .. link_mount_name .. "' already exists in link_mount table")
  end

  self:_exec(string.format([[
    INSERT INTO %s (link_name, knowledge_base, mount_path, description)
    VALUES (%s, %s, %s, %s)
  ]], mount_table,
      quote_literal(link_mount_name),
      quote_literal(knowledge_base),
      quote_literal(path),
      quote_literal(description)))

  self:_exec(string.format(
    "UPDATE %s SET has_link_mount = TRUE WHERE knowledge_base = %s AND path = %s",
    main_table, quote_literal(knowledge_base), quote_literal(path)))

  return knowledge_base, path
end

return KnowledgeBaseManager